libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -635,6 +680,45 @@ void CodeTransform::visitStatements(vector<Statement> const& _statements)
 	for (auto const& statement: _statements)
 	{
 		freeUnusedVariables();
+
+		if (!m_delayedReturnVariables.empty() && (
+			holds_alternative<VariableDeclaration>(statement) ||
+			holds_alternative<Leave>(statement) ||
+			holds_alternative<ForLoop>(statement) ||
+			holds_alternative<Block>(statement) ||
+			holds_alternative<Switch>(statement) ||
+			holds_alternative<If>(statement) ||
+			[&](){
+				ReferencesCounter referencesCounter{ReferencesCounter::CountWhat::OnlyVariables};
+				referencesCounter.visit(statement);
+				return ranges::any_of(
+					m_delayedReturnVariables,
+					[&](YulString _returnVariableName) { return referencesCounter.references().count(_returnVariableName); }
+				);
+			}()
+		))
+		{
+			for (YulString _returnVariableName: m_delayedReturnVariables)
+			{
+				auto* var = std::get_if<Scope::Variable>(m_scope->lookup(_returnVariableName));
+				yulAssert(var, "Return variable not in scope.");
+				m_context->variableStackHeights[var] = static_cast<size_t>(m_assembly.stackHeight());
+				// Preset stack slots for return variables to zero.
+				m_assembly.appendConstant(u256(0));
+				if (!m_unusedStackSlots.empty())
+				{
+					auto slot = static_cast<size_t>(*m_unusedStackSlots.begin());
+					m_unusedStackSlots.erase(m_unusedStackSlots.begin());
+					m_context->variableStackHeights[var] = slot;
+					if (size_t heightDiff = variableHeightDiff(*var, _returnVariableName, true))
+						m_assembly.appendInstruction(evmasm::swapInstruction(static_cast<unsigned>(heightDiff - 1)));
+					m_assembly.appendInstruction(evmasm::Instruction::POP);
+				}
+				++m_context->variableReferences[var];
```

Comment:
What is the function of this?

Reply:
Making sure return variables are always considered "used", preventing any funny business of the allocation algorithm freeing them or reallocating slots. Probably that wouldn't happen either way, but maybe better to be safe than sorry...

Reply:
Actually, the same is done in the VariableReferenceCounter up front anyways already, so I think I'll remove it here.